packages/x-data-grid-premium/src/hooks/features/pivoting/useGridPivoting.ts

```diff
@@ -0,0 +1,321 @@
+        columnGroupPath.push(String(colValue));
+        const groupId = columnGroupPath.join('-');
+
+        if (!columnGroupingModelLookup[groupId]) {
+          const columnGroup: GridColumnGroupingModel[number] = {
+            groupId,
+            headerName: String(colValue),
+            children: [],
+          };
+          columnGroupingModelLookup[groupId] = columnGroup;
+          if (depth === 0) {
+            columnGroupingModel.push(columnGroup);
+          } else {
+            const parentGroupId = columnGroupPath.slice(0, -1).join('-');
+            const parentGroup = columnGroupingModelLookup[parentGroupId];
+            parentGroup.children.push(columnGroup);
+          }
+        }
+
+        const isLastColumnGroupLevel = depth === pivotModel.columns.length - 1;
+
+        if (isLastColumnGroupLevel) {
+          pivotModel.values.forEach((pivotValue) => {
+            const valueField = pivotValue.field;
+            const valueKey = `${columnGroupPath.join('-')}-${valueField}`;
+            newRow[valueKey] = newRow[valueField];
+            delete newRow[valueField];
+          });
+        }
+      });
+
+      newRows.push(newRow);
+    });
+
+    sortColumnGroups(columnGroupingModel, pivotModel.columns);
+  }
+
+  function createColumns(columnGroups: GridColumnNode[], depth = 0) {
+    const initialColumnsLookup: Record<string, GridColDef | null> = {};
+
+    columnGroups.forEach((columnGroup) => {
+      if (isLeaf(columnGroup)) {
+        return;
+      }
+      const isLastColumnGroupLevel = depth === pivotModel.columns.length - 1;
+      if (isLastColumnGroupLevel) {
+        pivotModel.values.forEach((pivotValue) => {
+          if (typeof initialColumnsLookup[pivotValue.field] === 'undefined') {
+            initialColumnsLookup[pivotValue.field] =
+              columns.find((column) => column.field === pivotValue.field) || null;
+          }
+          const valueField = pivotValue.field;
+          const mapValueKey = `${columnGroup.groupId}-${valueField}`;
+          const column: GridColDef = {
+            field: mapValueKey,
+            headerName: String(valueField),
+            aggregable: true,
+            availableAggregationFunctions: [pivotValue.aggFunc],
+            ...getAttributesFromInitialColumn(pivotValue.field),
+          };
+          pivotColumns.push(column);
+          aggregationModel[mapValueKey] = pivotValue.aggFunc;
+          if (columnGroup) {
+            columnGroup.children.push({ field: mapValueKey });
+          }
+        });
+      } else {
+        createColumns(columnGroup.children, depth + 1);
+      }
+    });
+  }
+
+  createColumns(columnGroupingModel);
+
+  return {
+    rows: newRows,
+    columns: pivotColumns,
+    rowGroupingModel: pivotModel.rows,
+    aggregationModel,
+    getAggregationPosition: (groupNode) => (groupNode.depth === -1 ? 'footer' : 'inline'),
+    columnVisibilityModel,
+    columnGroupingModel,
```

Comment:
Column grouping appears to be the bottleneck when pivoting large datasets. It gets unacceptably slow if there are a lot of unique values that become column groups.
I'll see if I can make it faster.